### PR TITLE
Zeus QoL - Shelltracker + Notify on target damage

### PR DESCRIPTION
**Pull request by senshi-x**
These are some QoL changes and are not supposed to impact gameplay.
### Motivation/Goals: 
- Facilitate tracking of events during live gameplay for referees as well as possible observers/streamers.
- Minimize performance and network footprint (minimal remote calls)
- Remain flexible, allowing easy adaption/updating for possible future gameplay alterations

## Shelltracker
This module allows qualified players (currently only curators) to visualize artillery shells on the map.
### Features
- Shows a message whenever a player fires an artillery round.
- Visualizes shells in-flight on the map, as well as a lingering marker on the impact site including a radius marker for 95% effective blast range.
image

## Container Notify Damage
Adds a handleDamage EH that triggers a simple systemchat for Zeus players to inform them about containers taking damage. Might later also be used to automatically track target status, e.g. for mission end trigger.
Example:
`Blue 6 was hit by Senshi for 53.6883 % with Sh_82mm_AMOS . Current HP: 28.0927 %`

- When container is destroyed: Task and Marker get updated (for Zeus/Streamer only) to facilitate progress tracking.
-  Containers "remember" who landed the killing blow on them. Shows it for Zeus/Streamer, might later be used for fancy mission-end screen / stats
image

### CODE REVIEW

**Comment by NeutraleNull** [comment]
looking pretty good already. Could you try to revert the indent size? It changes the files all over the place. 
What branch are you working from? I see you have modified files that are not related to this PR?

**Comment by senshi-x** [comment]
Hence the "draft" status. Editorconfig will of course be changed back when ready for review, I just hate working with the 4 spaces in indent/block-heavy languages such as sqf (so much wasted real estate).

It's based off main branch, which is why the PR is targeting that as well. The only "other" file I changed right now is the lg postinit, because the eventhandler was interfering with my testing. Again, a thing that will change when I submit for review.

Same with description.ext changes, those are for debugging and won't be in here when it goes for review.

**Comment by b-mayr-1984** [comment]
Thx @senshi-x I will have a look at it in a few days. 🙂

**Comment by senshi-x** [comment]
I recommend waiting until I mark it as ready for review, so you don't have to wade through all the debug junk 😅 . Shouldn't take me long.

I feel like you guys have a different understanding of "draft" status in github than I do. 🙈 I use it to publish a description of "what I am working on", to prevent someone else from wasting energy on the same topic, and if that someone wants to dive into the same topic, they can look at WIP code if they really want to.

**Comment by b-mayr-1984** [comment]
> I recommend waiting until I mark it as ready for review, so you don't have to wade through all the debug junk 😅 . Shouldn't take me long.
> 
> I feel like you guys have a different understanding of "draft" status in github than I do. 🙈 I use it to publish a description of "what I am working on", to prevent someone else from wasting energy on the same topic, and if that someone wants to dive into the same topic, they can look at WIP code if they really want to.

Ah okay. Good to know. When I see a pull request I assume it is meant to be checked already.
I will wait then.

**Comment by senshi-x** [comment]
Ready for review.
@NeutraleNull Had to fix EH handling in lgshells that was  broken (idx of GetInMan instead of FiredEH was used to later attempt to remove FiredEH). Which obviously messes it up for everyone.

Sorry for some "useless" line changes being shown, mostly whitespace cleanups. Apparently your guys' editor doesn't enforce editorconfig's `trim_trailing_whitespace=true`?

File: scripts/lgshells/functions/fn_postInit.sqf
Comment (senshi-x) [inline scripts/lgshells/functions/fn_postInit.sqf:27]: This is the lgshells bug I had to fix. Decided to move to vehicle namespace to keep it localized, unless there's an explicit need for global var here that I don't see? @NeutraleNull

File: scripts/spot-randomizer/functions/fn_placeSupplyStation.sqf
Comment (senshi-x) [inline scripts/spot-randomizer/functions/fn_placeSupplyStation.sqf]: need to store these for onDamageTaken

File: scripts/lgshells/functions/fn_postInit.sqf
Comment (b-mayr-1984) [reply scripts/lgshells/functions/fn_postInit.sqf:27]: > This is the lgshells bug I had to fix. 

Which bug are you refering to?

File: scripts/spot-randomizer/functions/fn_randomizeSpots.sqf
Comment (b-mayr-1984) [inline scripts/spot-randomizer/functions/fn_randomizeSpots.sqf]: I intentionally left this out from the streamers map to not facilitate stream-sniping. 
Showing static targets on the streamers map for everyone to see is something that I wanted to avoid.

Same comment applies to line 54.

**Review by b-mayr-1984 — CHANGES_REQUESTED** [review]
Hi @senshi-x , 
Thanks for the cool feature. I think that people watching the stream will really enjoy this. 
And also for a Zeus it is good to know what's happening. 
The tracking feature really makes the game mode better. 🙂

Trying to understand what you changed to make this happen I really struggled though.
Here are the most upfront reasons:
- Your PR has a lot of changes. Many changed LOCs appear to be only cosmetic changes with no change to the runtime behaviour. 
- Your coding style is very different to what the other contributors to Shoot and Scoot have used (e.g. indentation depth but also density and style of code comments)


Both these make it really hard for me to review your code changes. And as the guy that probably ends up maintaining the code I should understand what's happening.

Please rework the PR to 
- only include changes in LOCs that actually have changes in behaviour
- use the same coding style like the contributors that came before this PR

Can't wait to have the feature on `main` then. 🤩

File: scripts/lgshells/functions/fn_postInit.sqf
Comment (senshi-x) [reply scripts/lgshells/functions/fn_postInit.sqf:27]: The one I mentioned in the discussion here https://github.com/gruppe-adler/Shoot_and_Scoot.Tanoa/pull/27#issuecomment-1637795166

**Comment by senshi-x** [comment]
Most of the "empty" changes result from using an IDE that honors editorconfig (by automatically enforcing/applying it), as also mentioned in https://github.com/gruppe-adler/Shoot_and_Scoot.Tanoa/pull/27#issuecomment-1637795166
Sure, those changes are only cosmetic and not functional changes, but I feel you can't really blame me for cleaning up code by accident. And I really don't feel like having to make code "dirtier" purely to reduce the amount of changed lines.

If the 2 space indent bothers you, I can change that, sure.
I do not know what you mean by "density and style of code comments", so not sure what you need changed there.

**Comment by b-mayr-1984** [comment]
> Most of the "empty" changes result from using an IDE that honors editorconfig (by automatically enforcing/applying it), as also mentioned in [#27 (comment)](https://github.com/gruppe-adler/Shoot_and_Scoot.Tanoa/pull/27#issuecomment-1637795166) 

I am aware of that. I read your comment there.

> Sure, those changes are only cosmetic and not functional changes, but I feel you can't really blame me for cleaning up code by accident. 

I don't want to blame you for anything. I just explain the reasons for me having trouble to review and get an understanding of your code.

> And I really don't feel like having to make code "dirtier" purely to reduce the amount of changed lines.

You don't have to make due with what you perceive as "dirty code". But putting behavioral changes in the same PR as refactoring changes makes it (unnecessarily) hard for a maintainer. Particularly if so many files and so many LOCs have been changed all at once.

> If the 2 space indent bothers you, I can change that, sure. 

That would really help, yes.

> I do not know what you mean by "density and style of code comments", so not sure what you need changed there.

I was a bit too general with my statement here. When I was writing this I was looking at [fn_onFired.sqf](https://github.com/gruppe-adler/Shoot_and_Scoot.Tanoa/blob/cafbd18da9899f6d5e561e5834c7ab188572f775/scripts/shelltracker/functions/fn_onFired.sqf). 
This file is very compressed. It does not make much use of elements to structure the code visually (e.g. aiding the eye by means of empty lines where functional blocks end or placing `[` and `{` etc. so that one can see where a closing bracket is that belongs to an opening one). 
If you look at other code files in this repository they look very different to this one and they make way more use of code comments.


To just add a little context:
- I tried to understand the changes of the PR by reviewing the code changes for more than an hour, which I wasn't successful at. 🙁 
- After giving up on the source code I then just jumped in-game and tried to test the new feature. Here I failed again because no marker would show up. 🙁 By pure chance I then found out that a player must fire the shell (an AI wouldn't work).

In my estimation this is what reviews are for. To see how others perceive changes. If others struggle there are probably reasons for it, and those reasons I am trying to present from my point of view.

**Comment by b-mayr-1984** [comment]
> > Most of the "empty" changes result from using an IDE that honors editorconfig (by automatically enforcing/applying it), as also mentioned in [#27 (comment)](https://github.com/gruppe-adler/Shoot_and_Scoot.Tanoa/pull/27#issuecomment-1637795166)
> 
> I am aware of that. I read your comment there.

Since this also caused confusion in the past (see [Diwako's PR](https://github.com/gruppe-adler/Shoot_and_Scoot.Tanoa/pull/1#issuecomment-1250051258)) and since my editor of choice (Notepad++ 👴) does not support it out of the box (but only with a plugin) I am trying to work with Visual Studio Code in the future. 

Maybe this makes things easier for future PRs.

**Comment by senshi-x** [comment]
How you want to code is entirely up to you, nobody likes being restricted in their methodology. But if you want people to adhere to a "codestyle", it makes sense to define a codestyle first. And editorconfig is a great way to do it in a way so people can still work with whatever IDE they prefer. Which is why I'm absolutely fine with moving to 4 space indents, even if I dislike them.
If your IDE does not enforce it, it still is not mandatory for you to change anything or QA your own code, but then you'll definitely have to keep dealing with "driveby" cleanups in other PRs such as mine.  ;)

**Comment by b-mayr-1984** [comment]
> How you want to code is entirely up to you, nobody likes being restricted in their methodology. 

Yes and no. I am used to work in larger development teams. This brings with it that individual freedom of methodology gets restricted by the need to effectivly work together. So I am fine with accepting some restrictions.

> But if you want people to adhere to a "codestyle", it makes sense to define a codestyle first. 

I don't really care about a specific codestyle. I care about being able to maintain the source code. This becomes easier if one project is not littered with different codestyles.

> And editorconfig is a great way to do it in a way so people can still work with whatever IDE they prefer. 

I agree. My own professional coding days lie more than 10 years in the past though. I've built my _coding muscle memory_ with editors that become increasingly obsolete with every day that passes. So in the spirit of teamwork I should move on rather than clinging to the old days. Even more so since with more people contributing here, my toolchain might become an increasing problem (e.g. by not supporting EditorConfig out of the box).

> Which is why I'm absolutely fine with moving to 4 space indents, even if I dislike them. If your IDE does not enforce it, it still is not mandatory for you to change anything or QA your own code,

👍 

> but then you'll definitely have to keep dealing with "driveby" cleanups in other PRs such as mine. ;)

Sorry, no! 🙅  Such a "driveby" action makes the review process to difficult. Delta-evaluation just becomes a mess. 
It is way better to separate functional changes and refactoring changes into separate change items (e.g. separate PRs). Just so that the reviewers/maintainers don't have to do so much reverse engineering of what you actually (functionally) changed.

Is there no way for you to revert the cosmetical changes back to how it was in the beginning?
You should have a way easier time doing that compared to me that hasn't actually implemented the changes.
I could easily overlook changes that are functional, but appear cosmetical to me, since I am not immersed in your thought processes.

**Comment by senshi-x** [comment]
I have changed the indents to 4 spaces to match editorconfig.

A suggestion to make reviewing this PR easier for you: In the "Files changed"-view of github, you can disable whitespace diffs. (small cog icon top left). Then you will only see "functional" changes. A similar function (ignore whitespace) should exist for most IDEs or git diff tools out there if you don't use github's webview.
image


A second suggestion: If preferred, we can just review this PR together sometime in TS/Discord. Might be easier especially if you have many questions and input regarding my code.

I will not undo any whitespace cleanup in this PR. That's a backwards approach.

I agree with you on the principle that refactor and functional changes should be separated. In this case, I didn't really refactor anything (renames), it's just whitespaces, so I feel it can be pragmatically included in this PR. If you disagree, the "proper" way to deal with this would be to simply do the whitespace cleanup in a separate PR and merge that first. Then the diff between main and this PR will no longer include those whitespace changes.

---

I also added another feature: tasks & markers get updated (for zeus/streamer only) once objectives are destroyed so they can easily track objective progress.

---
Code is ready for re-review.

**Comment by b-mayr-1984** [comment]
> I have changed the indents to 4 spaces to match editorconfig.

👍 

> A suggestion to make reviewing this PR easier for you: In the "Files changed"-view of github, you can disable whitespace diffs. (small cog icon top left). Then you will only see "functional" changes. A similar function (ignore whitespace) should exist for most IDEs or git diff tools out there if you don't use github's webview. 
>  image

Yes, using Visual Studio Code along with GitLens extension, as well as the setting that you mentioned, already makes my life way easier. I just need to get used to the new tools though.

> A second suggestion: If preferred, we can just review this PR together sometime in TS/Discord. Might be easier especially if you have many questions and input regarding my code.

That would be great, yes. 🙂 

> I will not undo any whitespace cleanup in this PR. That's a backwards approach.

Oh well, (reluctantly) accepted. 😐 

> I agree with you on the principle that refactor and functional changes should be separated. In this case, I didn't really refactor anything (renames), it's just whitespaces, so I feel it can be pragmatically included in this PR. If you disagree, the "proper" way to deal with this would be to simply do the whitespace cleanup in a separate PR and merge that first. Then the diff between main and this PR will no longer include those whitespace changes.

Exactly! 👍 

> I also added another feature: tasks & markers get updated (for zeus/streamer only) once objectives are destroyed so they can easily track objective progress.

I appreciate the feature. It implements something that was on my todo list and get's us closer to an **automated mission end**. 🙂 

From the review process point of view I have to admit though that it doesn't get easier with more/bigger changes. 
Having things separated and digestable in smaller bites would make it easier for maintainers.

> Code is ready for re-review.

I'll have a look at it. Thx 🙂

File: scripts/shelltracker/functions/fn_onFired.sqf
Comment (b-mayr-1984) [inline scripts/shelltracker/functions/fn_onFired.sqf]: @senshi-x replace occurences of `_radius` with `_diameter`.
Your multiplication by 2 and the rest of the code indicates to me that this value is a diameter rather than a radius.

File: scripts/shelltracker/functions/fn_onFired.sqf
Comment (b-mayr-1984) [inline scripts/shelltracker/functions/fn_onFired.sqf:44]: @senshi-x replace occurences of `_markerRadius` with `_markerHitRange`,  `_markerCircumference` or smilar.
Just as above this does not represent a radius. The marker visualizes an area of splash damage. This meaning should be reflected in the naming.

File: scripts/shelltracker/functions/fn_onFired.sqf
Comment (b-mayr-1984) [inline scripts/shelltracker/functions/fn_onFired.sqf]: @senshi-x add indentation to the commands in the `while`-loop.

File: scripts/spot-randomizer/functions/fn_randomizeSpots.sqf
Comment (b-mayr-1984) [inline scripts/spot-randomizer/functions/fn_randomizeSpots.sqf]: @senshi-x I am afraid that your change from `setMarkerText` to `setMarkerTextLocal` introduces a bug here.
This SQF file starts with `if (!isServer) exitWith {};`. Meaning no client will ever run it. So the changed marker text will not make it onto the map of the players.

Did you maybe use Search and Replace to generally over multiple files? Or was this change deliberate?

**Review by b-mayr-1984 — CHANGES_REQUESTED** [review]

File: scripts/shelltracker/functions/fn_onFired.sqf
Comment (b-mayr-1984) [reply scripts/shelltracker/functions/fn_onFired.sqf]: @senshi-x I understand the logic with the `indirectHitRange` now. Thanks for taking the time to explain it to me. 🙂 

The factor of 2 seems too big though, because the result looks like this:
image
But this area is way too big to realistically cause damage to. See screenshot below for scale of the village and the dirt that is thrown up:
image

I recommend lowering the factor down to `1.2` where the shell impact has lost only 50% of its damage potential.
image
(see https://community.bistudio.com/wiki/Arma_3:_Damage_Description#IndirectHit)

File: scripts/spot-randomizer/functions/fn_onDamageTaken.sqf
Comment (b-mayr-1984) [inline scripts/spot-randomizer/functions/fn_onDamageTaken.sqf:26]: @senshi-x I tested this on my machine 
- as a `Civilian` Zeus
- sitting in a BluFor MK6 mortar
- using it to kill a BluFor container
- all while hosting an MP session on my local machine
image

Never got any message or feedback on the map though. Feature does not appear to work. 🤔

File: scripts/spot-randomizer/functions/fn_onDamageTaken.sqf
Comment (b-mayr-1984) [inline scripts/spot-randomizer/functions/fn_onDamageTaken.sqf:39]: same as comment to line 26

**Review by b-mayr-1984 — CHANGES_REQUESTED** [review]
@senshi-x see my review comments.
- Shell tracking looks very nice. 🙂 
- Impact radius needs a little tweaking IMHO.
- Damage handling of the containers did not work during my tests.

File: scripts/spot-randomizer/functions/fn_placeSupplyStation.sqf
Comment (b-mayr-1984) [inline scripts/spot-randomizer/functions/fn_placeSupplyStation.sqf]: @senshi-x as discussed. This appears to be a code duplication that resulted from a merge. Please remove.

File: scripts/spot-randomizer/functions/fn_randomizeSpots.sqf
Comment (senshi-x) [reply scripts/spot-randomizer/functions/fn_randomizeSpots.sqf]: Fixed

File: scripts/shelltracker/functions/fn_onFired.sqf
Comment (senshi-x) [reply scripts/shelltracker/functions/fn_onFired.sqf]: Fixed

File: scripts/shelltracker/functions/fn_onFired.sqf
Comment (senshi-x) [reply scripts/shelltracker/functions/fn_onFired.sqf]: Changed as requested in 2fa6dfe611d133c91cd31115529f8be0cab5f9c5

File: scripts/spot-randomizer/functions/fn_onDamageTaken.sqf
Comment (senshi-x) [reply scripts/spot-randomizer/functions/fn_onDamageTaken.sqf:26]: Feature works (as requested by you ;) ) only on non-servers. `remoteExec -2` prevents it from being run on servers. If you player-host (not on dedicated server), you are the server, so it will not trigger.

https://github.com/gruppe-adler/Shoot_and_Scoot.Tanoa/blob/2fa6dfe611d133c91cd31115529f8be0cab5f9c5/scripts/spot-randomizer/functions/fn_placeSupplyStation.sqf#L125
If you want it to work on player-hosted server client as well, I recommend reverting it to remoteexec everywhere (remove -2). This has near-zero performance impact anyway, because the called onDamageTaken does a `hasInterface` check as the first action, and remoteexec is triggered by the server in the first place (so no added network traffic).

Also recommend to start testing MP missions on dedicated servers with multiple local clients. For any second or more local client the damage handler would work even on player-hosted servers.

Comment (b-mayr-1984) [reply scripts/spot-randomizer/functions/fn_onDamageTaken.sqf:26]: > If you want it to work on player-hosted server client as well, I recommend reverting it to remoteexec everywhere (remove -2). This has near-zero performance impact anyway, because the called onDamageTaken does a `hasInterface` check as the first action, and remoteexec is triggered by the server in the first place (so no added network traffic).

I changed it to `0` and tested it like this. Works fine now in my test setup. 🙂 

> Also recommend to start testing MP missions on dedicated servers with multiple local clients. For any second or more local client the damage handler would work even on player-hosted servers.

Sure. I will organize such a test session.
For the time being I think you did everything necessary that I will now merge the PR into `main`.

Thanks for the awesome set of new features. I love it!! 🤩

**Review by b-mayr-1984 — APPROVED** [review]

**Pull request merged**
